### PR TITLE
Reduce memory consumption in the input layer.

### PR DESCRIPTION
**Pull request by zheng-da**
*Issue #, if available:*

*Description of changes:*
To reduce memory consumption, we avoid computing the input embeddings in advance and saving them in distributed memory.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/model/gnn_encoder_base.py [hunk 1]
```diff
@@ -140,7 +139,10 @@ def dist_inference(g, gnn_encoder, node_feats, batch_size, fanout,
                     assert len(g.ntypes) == 1
                     output_nodes = {g.ntypes[0]: output_nodes}
 
-                h = {k: x[k][input_nodes[k]].to(device) for k in input_nodes.keys()}
+                if i == 0:
+                    h = get_input_embeds(blocks[0], input_nodes)
```
Comment (classicsong) [inline python/graphstorm/model/gnn_encoder_base.py:143]: Will this cause duplicated input feature computation?
I believe so.

File: python/graphstorm/model/gnn_encoder_base.py [hunk 2]
```diff
@@ -98,7 +98,6 @@ def dist_inference(g, gnn_encoder, node_feats, batch_size, fanout,
     dict of Tensor : the final GNN embeddings of all nodes.
     """
     device = gnn_encoder.device
-    x = node_feats
     with th.no_grad():
```
Comment (classicsong) [inline python/graphstorm/model/gnn_encoder_base.py:101]: The reason you want to compute input embedding within the for loop is that node_feats consumes many memory and we can not release it.

But if we do x = compute_node_input_embeddings(), x will automatically be released in L152.

File: python/graphstorm/model/gnn_encoder_base.py (hunk 1 above)
Comment (zheng-da) [reply python/graphstorm/model/gnn_encoder_base.py:143]: i will. it trades off computation for memory. if the input layer is very simple, such as MLP, this works fine. if the input layer is BERT, we probably don't want to do it.

File: python/graphstorm/model/gnn_encoder_base.py (hunk 2 above)
Comment (zheng-da) [reply python/graphstorm/model/gnn_encoder_base.py:101]: this approach requires to keep at least two embedding tensors of all nodes in memory.

Comment (classicsong) [reply python/graphstorm/model/gnn_encoder_base.py:101]: We always need two tensors (one for N-1 layer and one for N), right?

File: python/graphstorm/model/gnn_encoder_base.py (hunk 1 above)
Comment (classicsong) [reply python/graphstorm/model/gnn_encoder_base.py:143]: Shall we implement two logics to cover both use cases?

Comment (zheng-da) [reply python/graphstorm/model/gnn_encoder_base.py:143]: probably not here. I'll make some modifications in the place where `dist_inference` is called.

File: python/graphstorm/model/embed.py [hunk 3]
```diff
@@ -107,6 +107,19 @@ def get_sparse_params(self):
         # By default, there is no sparse_embeds
         return []
 
+    def require_cache_embed(self):
```
Comment (classicsong) [inline python/graphstorm/model/embed.py:110]: Can you also change GSLMNodeEncoderInputLayer and GSPureLMNodeInputLayer with require_cache_embed() returning True.

Comment (zheng-da) [reply python/graphstorm/model/embed.py:110]: they are changed.

**Review by classicsong — APPROVED** [review]
LGTM

**Review by classicsong — APPROVED** [review]
LGTM

**Pull request merged**